src/beatree/writeout.rs

```diff
@@ -264,18 +281,72 @@ impl BbnWriteOut {
 struct LnWriteOut {
     ln_fd: RawFd,
     ln_extend_file_sz: Option<u64>,
-    ln: Vec<()>,
+    ln: Vec<(PageNumber, Box<Page>)>,
+    ln_remaining: usize,
 }
 
 impl LnWriteOut {
-    fn run(&mut self) -> bool {
+    fn run(&mut self, io: &mut IoDmux) -> bool {
+        // NOTE: The current process begins by `ftruncating` the file to support writing
+        // all pages contained in `self.ln`.
+        // This is necessary because the pages created by the LeafStore fall into two categories:
+        // - pages that can be safely written to the file
+        // - pages that need the file to be extended because their page number is out of range
+        //
+        // The first category is currently waiting until `ftruncate` is performed before
+        // being written with the second group.
+        //
+        // A possible improvement could be to allow the LeafStore to write all pages
+        // in the first category immediately, keep a vector of pending pages,
+        // and have the writeout call `ftruncate` to finish issuing all pending pages.
```

Comment:
This is a good description, but once again, I am not sure what improvement this provides. As it is right now, the code that performs IO is concentrated in a single spot. Diffusing the responsibility is definitely a step back in this regard. Performance improvement might be a good reason to bite the bullet, we should measure, but even then now I am doubting if the improvement will be that good. I think likely ftruncate would just update the metadata of a file which is almost certainly cached, so that would be non-blocking. The actual write would be performed upon fsync.